[ENH] Move `latents` attribute to variable roles

daehyun99 opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2322

### List of changes to the codebase in this pull request
#### 1. `latents`
- [x] Delete `latents` argument and related logic in `DAG.add_node`, `DAG.add_nodes_from`.
- [x] Delete test code for testing `latents` argument.
- [x] Use the `GraphRolesMixin` to store the latent variables.
- [x] Add a property called `latents`.
- [x] Add test code for `latents` using `with_role` method of `DAG`, `PDAG`, `ADMG`, `AncestralBase`, ~~`UndirectedGraph`~~ classes.
- [ ] Update the documentation to show how to add a latent variable using keyword arguments.
#### 2. `weights`
- ~~Delete `weight` argument and related logic in `DAG.add_node`.~~
- ~~Delete test code for testing `weight` argument.~~
- ~~Remove the `add_node` and `add_nodes_from` methods from all classes.~~

daehyun99 committed:
Refactor: Delete `weight` argument and related logic in `DAG.add_node` (#2322)

ankurankan commented:
@daehyun99 I guess we found out where the `weight` argument was getting used. I think this should get resolved once you remove the `add_node` and `add_nodes_from` methods as `nx`'s methods should be able to accept the additional argument.

daehyun99 committed:
Revert, Refactor: Revert `weight` argument and Delete `latent` argument in `DAG.add_node`, `DAG.add_nodes_from` (#2322)

- Revert: Revert `weight` argument in `DAG.add_node`, `DAG.add_nodes_from`
- Refactor: Delete `latent` argument and related logic in `DAG.add_node`, `DAG.add_nodes_from`
- Refactor: Delete test code for testing `latents` argument with using `add_node`, `add_nodes_from` method.

daehyun99 commented:
@ankurankan  I have removed the `add_node` and `add_nodes_from` methods and tested it in my local environment, but the same error occurred.
Also, I found the following comment in the `DAG.add_nodes_from` method, but I need more time to understand what it means.
```py
    def add_nodes_from(
        self,
        nodes: Iterable[Hashable],
        weights: Optional[list[float] | tuple[float]] = None,
    ):
        """
        Add multiple nodes to the Graph.
        **The behviour of adding weights is different than in networkx.
        ...
        """
```
Therefore, I will perform the `latent`-related work first, and then perform the `weight`-related work.

daehyun99 committed:
Merge branch 'dev' into Refactor/2322

daehyun99 committed:
Refactor: Move `latents` attribute to variable roles (#2322)

- Added `latents` as an property of the `_GraphRolesMixin` class.
	This allows users to retrieve `latents` values via `model.latents`.
- As `latents` is now handled as roles, the initialization order has been changed in the `__init__` process of the `DAG` and `ADMG` classes.
- As `latents` is now handled as roles, we cannot specify a node that does not exist in the graph as `latents`.
	Consequently, the test code had to be modified.
- Deleted the `add_node` and `add_nodes_from` methods of the `ADMG` class.
- Deleted the test code that handles `latents` in the `add_node` and `add_nodes_from` methods of the `ADMG` class.

daehyun99 committed:
Fix: Solve the error in `SEM` when `latents` is changed to be handled as roles (#2322)

- In the case of `SEM`, information about the nodes is stored in `self.graph.nodes()` (whereas it is generally stored in `self.nodes()`).
	Therefore, I added an handling code for errors that occur when `latents` are handled as roles.
- As `latents` is now handled as roles, we cannot specify a node that does not exist in the graph as `latents`.
	Consequently, the test code had to be modified.

daehyun99 committed:
Style: Improve code readability `with_role` (#2322)

- Separated the logic for `SEM` graphs and non-`SEM` graphs.

daehyun99 committed:
Test: Add test code for `latents` using `with_role`, `without_role` method of `DAG` (#2322)

daehyun99 committed:
Test: Add test code for `latents` using `with_role`, `without_role` method of `PDAG` (#2322)

daehyun99 committed:
Test: Add test code for `latents` using `with_role`, `without_role` method of `ADMG` (#2322)

- `ADMG` doesn't have a `copy` method implemented.
- Therefore, `inplace=False` cannot be used for the `with_role` and `without_role` methods.
- Accordingly, I have written the test code only for the case where `inplace=True`.

daehyun99 committed:
Merge branch 'dev' into Refactor/2322

daehyun99 committed:
Merge branch 'Refactor/2322' of https://github.com/daehyun99/pgmpy into Refactor/2322

daehyun99 committed:
Feat: Add `_GraphRolesMixin` in `AncestralBase` class. (#2322)
Test: Add test code for `latents` using `with_role`, `without_role` method of `AncestralBase`

- `AncestralBase` doesn't have a `copy` method implemented.
- Therefore, `inplace=False` cannot be used for the `with_role` and `without_role` methods.
- Accordingly, I have written the test code only for the case where `inplace=True`.

daehyun99 committed:
Docs: Update the comments to show how to add `latents` (#2322)

daehyun99 commented:
@ankurankan Hi. Could you please do a code review only for the `latents` work?

While working on `latents`, I modified the code for the `AncestralBase` and `TestAncestralBase` classes. ([commit](https://github.com/pgmpy/pgmpy/pull/2338/commits/57741e1d43d8521d2cda4ed2cc2dd882890227b1))

This may cause conflicts with #2321.

Therefore, it might be confusing to do a code review after the `weights` work is also done.

Thanks.

ankurankan commented:
@daehyun99 Sorry for the delay. I think it looks good. One issue I see now is that because we have removed the `add_node` and `add_nodes_from` arguments, we can not add latent variables using these methods. Is there a workaround to this or using the `with_roles` method the only way?

ankurankan committed:
Merge branch 'dev' into Refactor/2322

ankurankan commented:
I have also resolved some merge conflicts on your branch, so please make sure to do a `git pull` before adding new commits.

daehyun99 committed:
Fix: `latents` should be a node declared in ebunch. (#2322)

daehyun99 commented:
> Is there a workaround to this or using the `with_roles` method the only way?

@ankurankan I have written the sample code below.

- How to add `latents` values to the declaration of a graph model
```py
# way1
>>> from pgmpy.base import DAG
>>> G = DAG(
...     ebunch=[("U", "X"), ("X", "M"), ("M", "Y"), ("U", "Y"), ("V", "M")],
...     latents={"U", "V"},
... )
>>> sorted(G.latents)
['U', 'V']

# way2
>>> from pgmpy.base import DAG
>>> G = DAG(
...     ebunch=[("U", "X"), ("X", "M"), ("M", "Y"), ("U", "Y"), ("V", "M")],
...     roles ={"latents": ("U", "V")},
... )
>>> sorted(G.latents)
['U', 'V']
```

- How to add `latents` values to a previously declared graph model
```py
# way1
>>> G.latents = "X"
>>> sorted(G.latents)
['U', 'V', 'X']

# way2
>>> G.with_role("latents", "X", inplace=True)
>>> sorted(G.latents)
['U', 'V', 'X']
```

- How to remove `latents` values from a graph model
```py
# only way
>>> G.without_role("latents", "X", inplace=True)
>>> sorted(G.latents)
['U', 'V']
```

daehyun99 commented:
@ankurankan Ready to merge.

ankurankan commented on pgmpy/base/_mixin_roles.py:
```diff
@@ -149,3 +156,31 @@ def is_valid_causal_structure(self) -> bool:
+
+    @latents.setter
+    def latents(self, variables):
+        self.with_role(role="latents", variables=variables, inplace=True)
```
I don't find this behavior for a setter method to be intuitive. I think the setter method should replace the existing latents with the ones provided in `variables`. For the general modification operations, we would still have the `with` and `without`  role method. What do you think?

daehyun99 replied:
@ankurankan 
Yes. I also think the behavior of the setter is not intuitive. 
I've modified the code to replace the existing `latents` value.

```py
    @latents.setter
    def latents(self, variables):
        """
        Replace the `latents` nodes.

        Parameters
        ----------
        variables: set of nodes (default: empty set)
            A set of latent variables in the graph. These are not observed
            variables but are used to represent unobserved confounding or
            other latent structures.
        """
        if self.has_role("latents"):
            self.without_role(
                role="latents", variables=self.get_role("latents"), inplace=True
            )
        self.with_role(role="latents", variables=variables, inplace=True)
```

ankurankan commented on pgmpy/tests/test_base/test_ADMG.py:
```diff
@@ -56,6 +57,50 @@ def test_initialization_with_roles(self):
+        with pytest.raises(ValueError, match="Variable 'G' not found in the graph."):
+            admg.with_role(role="latents", variables="G", inplace=True)
+
+    def test_latnets_without_role(self):
```
A small typo: `_latnets_` -> `_latents_`

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -73,79 +73,36 @@ def test_add_node_string(self):
-
-        self.graph = DAG()
-        self.graph.add_nodes_from([1, 2, 3, 4], latent=[True, False, False, False])
+        # TODO: What needs to be resolved before merging PR #2338
```
Thanks for catching this. Could you add some assert statements here, please?

daehyun99 replied:
It has been applied.
Additionally, I added assert statements for some tests.

ankurankan reviewed (changes requested):
@daehyun99 Thanks a lot for the PR, it mostly looks good to me. I have added a few very minor comments. Could you please have a look?

daehyun99 committed:
Merge branch 'dev' into Refactor/2322

daehyun99 committed:
Feat: Update `@latents.setter` (#2322)

- Apply Reviews #2338

ankurankan reviewed (approved):
@daehyun99 Looks good to me. Thanks a lot for the PR.
